src/category_theory/differential_object.lean

```diff
@@ -0,0 +1,124 @@
+import category_theory.limits.shapes.zero
+import category_theory.shift
+
+open category_theory.limits
+
+universes v u
+
+namespace category_theory
+
+variables (C : Type u) [𝒞 : category.{v} C]
+include 𝒞
+
+variables [has_zero_morphisms.{v} C] [has_shift.{v} C]
+
+/--
+A differential object in a category with zero morphisms and a shift is
+an object `X` equipped with
+a morphism `d : X ⟶ X[1]`, so `d^2 = 0`.
+-/
+@[nolint has_inhabited_instance]
+structure differential_object :=
+(X : C)
+(d : X ⟶ X[1])
+(d_squared' : d ≫ d[[1]] = 0 . obviously)
+
+restate_axiom differential_object.d_squared'
+attribute [simp] differential_object.d_squared
+
+variables {C}
+
+namespace differential_object
+
+/--
+A morphism of differential objects is a morphism commuting with the differentials.
+-/
+@[ext, nolint has_inhabited_instance]
+structure hom (X Y : differential_object.{v} C) :=
+(f : X.X ⟶ Y.X)
+(comm' : f ≫ Y.d = X.d ≫ f[[1]] . obviously)
+
+restate_axiom hom.comm'
+
+namespace hom
+
+/-- The identity morphism of a differential object. -/
+@[simps]
+def id (X : differential_object.{v} C) : hom X X :=
+{ f := 𝟙 X.X }
+
+/-- The composition of morphisms of differential objects. -/
+@[simps]
+def comp {X Y Z : differential_object.{v} C} (f : hom X Y) (g : hom Y Z) : hom X Z :=
+{ f := f.f ≫ g.f,
+  comm' := by rw [category.assoc, g.comm, ←category.assoc, f.comm, category.assoc, functor.map_comp], }
+
+end hom
+
+instance category_of_differential_objects : category.{v} (differential_object.{v} C) :=
+{ hom := hom,
+  id := hom.id,
+  comp := λ X Y Z f g, hom.comp f g, }
+
+@[simp]
+lemma id_f (X : differential_object.{v} C) : ((𝟙 X) : hom X X).f = 𝟙 (X.X) := rfl
```

Comment:
```suggestion
lemma id_f (X : differential_object.{v} C) : ((𝟙 X) : X ⟶ X).f = 𝟙 (X.X) := rfl
```

Reply:
No, this was intentional, so Lean can understand what the `.f` is doing.